pkg/apis/pipeline/v1beta1/param_types.go

```diff
@@ -160,6 +174,14 @@ func NewArrayOrString(value string, values ...string) *ArrayOrString {
 	}
 }
 
+// NewObject creates an ArrayOrString of type ParamTypeObject using the provided key-value pairs
```

Comment:
Nice job putting into thoughts about how to do them uniformly. With the considerations that ArrayOrString will be renamed to something else, I'm proposing to call this new struct as StructuredObject, to distinguish this struct, which could be containing strings, arrays or maps, from the "object" (dictionary) concept in json schema.

With this proposal, and because Go doesn't support function overloading, I think it makes sense to actually have three functions, NewStructuredObjectInString, NewStructuredObjectInObject, and NewStructuredObjectInObject. Pinging @vdemeester @pritidesai for opinions.

Reply:
Great point! Yeah we definitely need to rename the struct `ArrayOrString` to something more general. Thanks for calling this out. The word `object` in `StructuredObject ` might be confusing with the actual `object` type that we will support. How about `StructuredParamValue`?

In terms of splitting `NewArrayOrString` function into three, I don't see any issue with that because this function is mainly used in test and just one place in the implementation code - `resolveResultRef` [function](https://github.com/tektoncd/pipeline/blob/2b8538214169ec1c3af14221f44ea59e6b1b81f7/pkg/reconciler/pipelinerun/resources/resultrefresolution.go#L158). We just need to be very careful with refactor because this function is extensively used in unit tests.

Reply:
> 

Ah this will be shared with results as well later

Reply:
> > 
> 
> Ah this will be shared with results as well later

yes, result will use this struct `ArrayOrString ` as well. cc: @Yongxuanzhang

Reply:
> > > 
> > 
> > 
> > Ah this will be shared with results as well later
> 
> yes, result will use this struct `ArrayOrString ` as well. cc: @Yongxuanzhang

yes it is already in my next pr, thanks for cc me!

Reply:
In today's Pipeline WG meeting, @pritidesai proposed to create a new struct which embeds `ArrayOrString` and then includes `object` type to avoid breaking changes IICU.

My thoughts:
- If we create the new struct and keep `ArrayOrString` there, we still cannot avoid breaking changes unfortunately. For instance, we have to change the fields which were of `ArrayOrString` type i.e. `ParamSpec`'s [default](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/param_types.go#L52) and `Param`'s [value](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/param_types.go#L85).
- Renaming `ArrayOrString` might make more sense to me too. 

Can you please share more thoughts regarding this here? @pritidesai

Reply:
Just agreed on that we'll solve this in another pr. @ywluogg @Yongxuanzhang

Reply:
Thanks for updating this. Can you add a TODO for it? And you can mark this as resolved